File: python/src/symphony/tracker/linear.py
import logging
from datetime import datetime, timezone
from typing import Any

from symphony.tracker.base import TrackerAdapter
from symphony.models import Issue, BlockerRef
from symphony.exceptions import (
    MissingTrackerApiKey, LinearApiRequest, LinearGraphQLErrors,
    LinearUnknownPayload, LinearMissingEndCursor,
)

logger = logging.getLogger(__name__)

CANDIDATE_QUERY = """
query CandidateIssues($projectSlug: String!, $activeStates: [String!]!, $after: String, $pageSize: Int!) {
  projects(filter: { slugId: { eq: $projectSlug } }) {
    nodes {
      id
      issues(first: $pageSize, after: $after, filter: { state: { name: { in: $activeStates } } }) {
        nodes {
          id
          identifier
          title
          description
          priority
          state { name }
          branchName
          url
          labels { nodes { name } }
          children { nodes { id identifier state { name } } }
          createdAt
          updatedAt
        }
        pageInfo {
          hasNextPage
          endCursor
        }
      }
    }
  }
}
"""

STATE_REFRESH_QUERY = """
query IssueStates($ids: [ID!]!) {
  issues(filter: { id: { in: $ids } }) {
    nodes {
      id
      identifier
      title
      description
      priority
      state { name }
      url
      labels { nodes { name } }
      createdAt
      updatedAt
    }
  }
}
"""

TERMINAL_QUERY = """
query TerminalIssues($projectSlug: String!, $terminalStates: [String!]!) {
  projects(filter: { slugId: { eq: $projectSlug } }) {
    nodes {
      id
      issues(filter: { state: { name: { in: $terminalStates } } }) {
        nodes {
          id
          identifier
        }
      }
    }
  }
}
"""


class LinearTracker(TrackerAdapter):
    def __init__(
        self,
        api_key: str,
        project_slug: str,
        endpoint: str = "https://api.linear.app/graphql",
        active_states: list[str] | None = None,
        terminal_states: list[str] | None = None,
        page_size: int = 50,
        timeout: int = 30,
    ):
        if not api_key:
            raise MissingTrackerApiKey("Linear API key is required")
        self.api_key = api_key
        self.project_slug = project_slug
        self.endpoint = endpoint
        self.active_states = active_states or ["Todo", "In Progress"]
        self.terminal_states = terminal_states or [
            "Closed", "Cancelled", "Canceled", "Duplicate", "Done",
        ]
        self.page_size = page_size
        self.timeout = timeout

    def fetch_candidate_issues(self) -> list[Issue]:
        issues: list[Issue] = []
        after: str | None = None
        while True:
            query, variables = self._build_candidate_query(after)
            data = self._execute(query, variables)
            project_nodes = self._navigate(data, ["data", "projects", "nodes"])
            if not project_nodes:
                break
            issue_connection = project_nodes[0].get("issues", {})
            page_issues = self._parse_issue_nodes(issue_connection.get("nodes", []))
            issues.extend(page_issues)
            page_info = issue_connection.get("pageInfo", {})
            if not page_info.get("hasNextPage"):
                break
            after = page_info.get("endCursor")
            if not after:
                raise LinearMissingEndCursor("Missing endCursor for pagination")
        return issues

    def fetch_issues_by_states(self, state_names: list[str]) -> list[Issue]:
        if not state_names:
            return []
        variables = {
            "projectSlug": self.project_slug,
            "terminalStates": state_names,
        }
        data = self._execute(TERMINAL_QUERY, variables)
        project_nodes = self._navigate(data, ["data", "projects", "nodes"])
        if not project_nodes:
            return []
        issue_nodes = project_nodes[0].get("issues", {}).get("nodes", [])
        return [
            Issue(id=n["id"], identifier=n["identifier"], title="", state="")
            for n in issue_nodes
        ]

    def fetch_issue_states_by_ids(self, issue_ids: list[str]) -> list[Issue]:
        if not issue_ids:
            return []
        variables = {"ids": issue_ids}
        data = self._execute(STATE_REFRESH_QUERY, variables)
        nodes = self._navigate(data, ["data", "issues", "nodes"]) or []
        return [self._normalize_issue(n) for n in nodes]

    def _build_candidate_query(self, after: str | None) -> tuple[str, dict]:
        variables = {
            "projectSlug": self.project_slug,
            "activeStates": self.active_states,
            "after": after,
            "pageSize": self.page_size,
        }
        return CANDIDATE_QUERY, variables

    def _execute(self, query: str, variables: dict[str, Any]) -> dict:
        """Execute a GraphQL query synchronously using aiohttp."""
        import asyncio
        try:
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
            try:
                result = loop.run_until_complete(self._async_execute(query, variables))
                return result
            finally:
                loop.close()
        except LinearGraphQLErrors:
            raise
        except LinearApiRequest:
            raise
        except Exception as e:
            raise LinearApiRequest(f"Linear API request failed: {e}")

    async def _async_execute(self, query: str, variables: dict[str, Any]) -> dict:
        import aiohttp
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }
        payload = {"query": query, "variables": variables}
        async with aiohttp.ClientSession() as session:
            async with session.post(
                self.endpoint,
                json=payload,
                headers=headers,
                timeout=aiohttp.ClientTimeout(total=self.timeout),
            ) as resp:
                if resp.status != 200:
                    text = await resp.text()
                    raise LinearApiRequest(
                        f"Linear API returned status {resp.status}: {text[:500]}"
                    )
                data = await resp.json()

        if "errors" in data and data["errors"]:
            raise LinearGraphQLErrors(f"GraphQL errors: {data['errors']}")

        return data

    def _navigate(self, data: dict, path: list[str]) -> Any:
        current = data
        for key in path:
            if not isinstance(current, dict):
                return None
            current = current.get(key)
        return current

    def _parse_issue_nodes(self, nodes: list[dict]) -> list[Issue]:
        return [self._normalize_issue(n) for n in nodes]

    def _normalize_issue(self, node: dict) -> Issue:
        state_name = ""
        if isinstance(node.get("state"), dict):
            state_name = node["state"].get("name", "")

        labels = []
        if isinstance(node.get("labels"), dict):
            label_nodes = node["labels"].get("nodes", [])
            labels = [l.get("name", "") for l in label_nodes if isinstance(l, dict)]

        blockers = []
        if isinstance(node.get("children"), dict):
            for child in node["children"].get("nodes", []):
                if isinstance(child, dict):
                    blockers.append(BlockerRef(
                        id=child.get("id"),
                        identifier=child.get("identifier"),
                        state=child.get("state", {}).get("name") if isinstance(child.get("state"), dict) else None,
                    ))

        priority = node.get("priority")
        if not isinstance(priority, int):
            priority = None

        created = None
        if node.get("createdAt"):
            try:
                created = datetime.fromisoformat(node["createdAt"].replace("Z", "+00:00"))
            except (ValueError, TypeError):
                pass

        updated = None
        if node.get("updatedAt"):
            try:
                updated = datetime.fromisoformat(node["updatedAt"].replace("Z", "+00:00"))
            except (ValueError, TypeError):
                pass

        return Issue(
            id=node.get("id", ""),
            identifier=node.get("identifier", ""),
            title=node.get("title", ""),
            description=node.get("description"),
            priority=priority,
            state=state_name,
            branch_name=node.get("branchName"),
            url=node.get("url"),
            labels=labels,
            blocked_by=blockers,
            created_at=created,
            updated_at=updated,
        )

File: typescript/src/tracker/linear.ts
import type { Issue, BlockerRef } from '../models'
import type { TrackerAdapter } from './base'

interface LinearConfig {
  endpoint: string
  apiKey: string
  projectSlug: string
  activeStates: string[]
  terminalStates: string[]
}

interface LinearIssueNode {
  id: string
  identifier: string
  title: string
  description?: string | null
  priority?: number | null
  branchName?: string | null
  url?: string | null
  labels?: { nodes?: Array<{ name: string }> }
  state?: { name: string }
  createdAt?: string
  updatedAt?: string
  children?: { nodes?: Array<{ id: string; identifier: string; state?: { name: string } }> }
}

function normalizeIssue(node: LinearIssueNode): Issue {
  const blockers: BlockerRef[] = (node.children?.nodes ?? []).map((c) => ({
    id: c.id ?? null,
    identifier: c.identifier ?? null,
    state: c.state?.name ?? null,
  }))

  return {
    id: node.id,
    identifier: node.identifier,
    title: node.title,
    state: node.state?.name ?? 'Unknown',
    description: node.description ?? null,
    priority: typeof node.priority === 'number' ? node.priority : null,
    branchName: node.branchName ?? null,
    url: node.url ?? null,
    labels: (node.labels?.nodes ?? []).map((l) => l.name.toLowerCase()),
    blockedBy: blockers,
    createdAt: node.createdAt ? new Date(node.createdAt) : null,
    updatedAt: node.updatedAt ? new Date(node.updatedAt) : null,
  }
}

export class LinearTracker implements TrackerAdapter {
  constructor(private config: LinearConfig) {}

  async fetchCandidateIssues(): Promise<Issue[]> {
    const allNodes: LinearIssueNode[] = []
    let cursor: string | null = null
    const pageSize = 50

    while (true) {
      const query = `query Candidates($projectSlug: String!, $activeStates: [String!]!, $after: String, $first: Int!) {
        issues(filter: { project: { slugId: { eq: $projectSlug } }, state: { name: { in: $activeStates } } }, first: $first, after: $after) {
          nodes { id identifier title description priority branchName url
            labels { nodes { name } }
            state { name }
            createdAt updatedAt
            children { nodes { id identifier state { name } } } }
          pageInfo { hasNextPage endCursor }
        }
      }`
      const data = await this.graphql<{ issues: { nodes: LinearIssueNode[]; pageInfo: { hasNextPage: boolean; endCursor: string | null } } }>(query, {
        projectSlug: this.config.projectSlug,
        activeStates: this.config.activeStates,
        after: cursor,
        first: pageSize,
      })
      if (!data?.issues) break
      allNodes.push(...data.issues.nodes)
      if (!data.issues.pageInfo.hasNextPage) break
      cursor = data.issues.pageInfo.endCursor
      if (!cursor) break
    }

    return allNodes.map(normalizeIssue)
  }

  async fetchIssuesByStates(stateNames: string[]): Promise<Issue[]> {
    const query = `query ByStates($stateNames: [String!]!) {
      issues(filter: { state: { name: { in: $stateNames } } }, first: 50) {
        nodes { id identifier title state { name } }
      }
    }`
    const data = await this.graphql<{ issues: { nodes: LinearIssueNode[] } }>(query, { stateNames })
    return (data?.issues?.nodes ?? []).map(normalizeIssue)
  }

  async fetchIssueStatesByIds(issueIds: string[]): Promise<Issue[]> {
    const query = `query ByIds($ids: [ID!]!) {
      issues(filter: { id: { in: $ids } }, first: 50) {
        nodes { id identifier title state { name } }
      }
    }`
    const data = await this.graphql<{ issues: { nodes: LinearIssueNode[] } }>(query, { ids: issueIds })
    return (data?.issues?.nodes ?? []).map(normalizeIssue)
  }

  private async graphql<T>(query: string, variables: Record<string, unknown>): Promise<T | null> {
    const response = await fetch(this.config.endpoint, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: this.config.apiKey },
      body: JSON.stringify({ query, variables }),
      signal: AbortSignal.timeout(30000),
    })
    if (!response.ok) throw new Error(`Linear API returned ${response.status}`)
    const body = (await response.json()) as { data?: T; errors?: Array<{ message: string }> }
    if (body.errors) throw new Error(`Linear GraphQL errors: ${body.errors.map((e) => e.message).join(', ')}`)
    return body.data ?? null
  }
}
